### PR TITLE
Support for RFC 3161 cryptographic timestamps with RFC 5816 additions

### LINKED ISSUE #24524: Support RFC 5816 and RFC 3161 trusted timestamping
API proposal:

```C#
namespace System.Security.Cryptography.Pkcs
{
    public sealed partial class Rfc3161TimestampRequest
    {
        private Rfc3161TimestampRequest() { }
        public int Version => throw null;
        public ReadOnlyMemory<byte> GetMessageHash() => throw null;
        public Oid HashAlgorithmId => throw null;
        public Oid RequestedPolicyId => throw null;
        public bool RequestSignerCertificate => throw null;
        public ReadOnlyMemory<byte>? GetNonce() => throw null;
        public bool HasExtensions => throw null;
        public X509ExtensionCollection GetExtensions() => throw null;
        public byte[] Encode() => throw null;
        public bool TryEncode(Span<byte> destination, out int bytesWritten) => throw null;
        public Rfc3161TimestampToken ProcessResponse(
                ReadOnlyMemory<byte> responseBytes, out int bytesConsumed) => throw null;
        public static Rfc3161TimestampRequest CreateFromData(
                ReadOnlySpan<byte> data, HashAlgorithmName hashAlgorithm, Oid requestedPolicyId = null, ReadOnlyMemory<byte>? nonce = null, bool requestSignerCertificates = false, X509ExtensionCollection extensions = null) => throw null;
        public static Rfc3161TimestampRequest CreateFromHash(
                ReadOnlyMemory<byte> hash, HashAlgorithmName hashAlgorithm, Oid requestedPolicyId = null, ReadOnlyMemory<byte>? nonce = null, bool requestSignerCertificates = false, X509ExtensionCollection extensions = null) => throw null;
        public static Rfc3161TimestampRequest CreateFromHash(
                ReadOnlyMemory<byte> hash, Oid hashAlgorithmId, Oid requestedPolicyId = null, ReadOnlyMemory<byte>? nonce = null, bool requestSignerCertificates = false, X509ExtensionCollection extensions = null) => throw null;
        public static Rfc3161TimestampRequest CreateFromSignerInfo(
                SignerInfo signerInfo, HashAlgorithmName hashAlgorithm, Oid requestedPolicyId = null, ReadOnlyMemory<byte>? nonce = null, bool requestSignerCertificates = false, X509ExtensionCollection extensions = null) => throw null;
        public static bool TryDecode(
                ReadOnlyMemory<byte> encodedBytes, out Rfc3161TimestampRequest request, out int bytesConsumed) => throw null;
    }
    public sealed partial class Rfc3161TimestampToken
    {
        private Rfc3161TimestampToken() { }
        public Rfc3161TimestampTokenInfo TokenInfo => throw null;
        public SignedCms AsSignedCms() => throw null;
        public bool VerifySignatureForHash(
                ReadOnlySpan<byte> hash, HashAlgorithmName hashAlgorithm, out X509Certificate2 signerCertificate, X509Certificate2Collection extraCandidates = null) => throw null;
        public bool VerifySignatureForHash(
                ReadOnlySpan<byte> hash, Oid hashAlgorithmId, out X509Certificate2 signerCertificate, X509Certificate2Collection extraCandidates = null) => throw null;
        public bool VerifySignatureForData(
                ReadOnlySpan<byte> data, out X509Certificate2 signerCertificate, X509Certificate2Collection extraCandidates = null) => throw null;
        public bool VerifySignatureForSignerInfo(
                SignerInfo signerInfo, out X509Certificate2 signerCertificate, X509Certificate2Collection extraCandidates = null) => throw null;
        public static bool TryDecode(
                ReadOnlyMemory<byte> encodedBytes, out Rfc3161TimestampToken token, out int bytesConsumed) => throw null;
    }
    public sealed partial class Rfc3161TimestampTokenInfo
    {
        public Rfc3161TimestampTokenInfo(
                Oid policyId, Oid hashAlgorithmId, ReadOnlyMemory<byte> messageHash, ReadOnlyMemory<byte> serialNumber, DateTimeOffset timestamp, long? accuracyInMicroseconds=null, bool isOrdering=false, ReadOnlyMemory<byte>? nonce=null, ReadOnlyMemory<byte>? timestampAuthorityName=null, X509ExtensionCollection extensions =null) { throw null; }
        public int Version => throw null;
        public Oid PolicyId=> throw null;
        public Oid HashAlgorithmId => throw null;
        public ReadOnlyMemory<byte> GetMessageHash() { throw null; }
        public ReadOnlyMemory<byte> GetSerialNumber() { throw null; }
        public DateTimeOffset Timestamp => throw null;
        public long? AccuracyInMicroseconds => throw null;
        public bool IsOrdering => throw null;
        public ReadOnlyMemory<byte>? GetNonce() { throw null; }
        public ReadOnlyMemory<byte>? GetTimestampAuthorityName() { throw null; }
        public bool HasExtensions => throw null;
        public X509ExtensionCollection GetExtensions() { throw null; }
        public byte[] Encode() => throw null;
        public bool TryEncode(Span<byte> destination, out int bytesWritten) => throw null;
        public static bool TryDecode(
                ReadOnlyMemory<byte> encodedBytes, out Rfc3161TimestampTokenInfo timestampTokenInfo, out int bytesConsumed) { throw null; }
    }
}
```

https://tools.ietf.org/html/rfc5816#section-1

https://tools.ietf.org/html/rfc3161

https://en.wikipedia.org/wiki/Trusted_timestamping

### PR DESCRIPTION
**Pull request by bartonjs**
This change adds API to inspect, request, and create timestamp tokens in accordance with IETF RFC 3161 (Internet X.509 Public Key Infrastructure Time-Stamp Protocol (TSP)).  The validation layer includes support for IETF RFC 5816 (ESSCertIDv2 Update for RFC 3161).

Unlike the Win32 CryptRetrieveTimeStamp, this API does not provide "Time-Stamp Protocol via HTTP" functionality.  Instead it provides enough of the processing logic to enable a client to do the HTTP(S) request using whatever preferred technology they desire, or to implement any of the other protocols (email, file, sockets, or a custom solution) as their application sees fit.

Rfc3161TimestampRequest.CreateForSignerInfo and Rfc3161TimestampToken.VerifyForSignerInfo provide an accelerated path for IETF RFC 3161 Appendix A (Signature Time-stamp attribute using CMS).

References:
* https://tools.ietf.org/html/rfc3161
* https://tools.ietf.org/html/rfc5816

Fixes #24524.

### CODE REVIEW

File: src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx [hunk 1]
```diff
@@ -244,6 +244,21 @@
...
+    <value>The response from the timestamping server did not match the request nonce.</value>
+  </data>
+  <data name="Cryptography_TimestampReq_BadResponse" xml:space="preserve">
+    <value>The response from the timestamping server was not understood.</value>
+  </data>
+  <data name="Cryptography_TimestampReq_Failure" xml:space="preserve">
+    <value>The timestamping server did not grant the request. The request status is '{0}' with failure info '{1}'.</value>
+  </data>
+  <data name="Cryptography_TimestampReq_NoCertFound" xml:space="preserve">
+    <value>The timestamping request requested the TSA certificate in the response, but it was not found.</value>
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx]: should "request requested" actually be "request required"

File: src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx [hunk 2]
```diff
@@ -244,6 +244,21 @@
...
+  </data>
+  <data name="Cryptography_TimestampReq_UnexpectedCertFound" xml:space="preserve">
+    <value>The timestamping request requested the TSA certificate not be included in the response, but certificates were present.</value>
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx]: Same thing here - should "request requested" actually be "request required"

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/GeneralName.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        internal ReadOnlyMemory<byte>? TeletexString;
+
+        [PrintableString]
+        internal string PrintableString;
+
+        [ExpectedTag(TagClass.Universal, (int)UniversalTagNumber.UniversalString)]
+        internal ReadOnlyMemory<byte>? UniversalString;
+
+        [UTF8String]
+        internal string Utf8String;
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/GeneralName.cs]: why isn't this a byte[]

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161Accuracy.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        [ExpectedTag(0)]
+        [OptionalValue]
+        internal int? Millis;
+
+        [ExpectedTag(1)]
+        [OptionalValue]
+        internal int? Micros;
+
+        internal Rfc3161Accuracy(long totalMicros)
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161Accuracy.cs]: nit: zero is valid value for accuracy in the spec;

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/X509ExtensionAsn.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+using System.Security.Cryptography.X509Certificates;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs.Asn1
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct X509ExtensionAsn
+    {
+        [ObjectIdentifier]
+        internal string ExtnId;
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/X509ExtensionAsn.cs]: Do you use the Extn abbreviation anywhere else?  I'd personally rather this just be "Id"

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/X509ExtensionAsn.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
+        [DefaultValue(0x01, 0x01, 0x00)]
+#pragma warning restore CS3016 // Arrays as attribute arguments is not CLS-compliant
+        internal bool Critical;
+
+        [OctetString]
+        internal ReadOnlyMemory<byte> ExtnValue;
```
Comment (joshfree) [inline src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/X509ExtensionAsn.cs]: same for this - why not just use Value

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/GeneralName.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/GeneralName.cs]: Why would it be?

File: src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx (hunk 2 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx]: I can go either way.  Technically the request merely requested it.  The RFC then says that the response should only be accepted if this request was fulfilled.

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161Accuracy.cs (hunk 4 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161Accuracy.cs]: And, apparently I had a test for reading zero successfully, but not for writing it.  (It was written by hand-rolling the tokenInfo DER).  Good catch, fixed.

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampRequest.cs [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+            status = Rfc3161RequestResponseStatus.Unknown;
+            token = null;
+
+            Rfc3161TimeStampResp resp;
+
+            try
+            {
+                resp = AsnSerializer.Deserialize<Rfc3161TimeStampResp>(source, AsnEncodingRules.DER, out bytesConsumed);
+            }
+            catch (CryptographicException)
```
Comment (GrabYourPitchforks) [inline src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampRequest.cs]: Nit: `catch (CryptographicException) when (!shouldThrow) { ... }` is generally preferred to rethrowing exceptions. This makes debugging a bit easier. See https://www.thomaslevesque.com/2015/06/21/exception-filters-in-c-6/ for more info.

File: src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161TimeStampReq.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        public MessageImprint MessageImprint;
+
+        [OptionalValue]
+        public Oid ReqPolicy;
+
+        [OptionalValue]
+        [Integer]
+        public ReadOnlyMemory<byte>? Nonce;
+
+#pragma warning disable CS3016
```
[the remark this answers is not in the document]
Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Asn1/Rfc3161TimeStampReq.cs]: Some people are quick to suppress a warning in a project, and I guess I balance them out by hesitating.  Next time I have to suppress one of these I'll probably do it at the project level.

File: src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx (hunk 2 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/Resources/Strings.resx]: "Either way" + one more nudge => messages changed.

**Comment by bartonjs** [comment]
@dotnet-bot Test Linux arm Release Build please (build "triggered" but hasn't assigned in 3 hours)
Test Tizen armel Debug Build please (build "triggered" but hasn't assigned in 3 hours)

**Comment by bartonjs** [comment]
@dotnet/dncenghot The Linux arm Release and Tizen armel Debug legs have been "Triggered" for 3.5 hours.  I've seen other PRs successfully report a position in the queue and complete, but this one hasn't updated.  Asking for these specific legs to be (re-)tested didn't seem to change the state of affairs, and I'm resisting the big restart trigger phrase.

If there's a surgical manner of fixing this I'd like to get to green.  And, of course, we'd ideally never get into this state...

**Comment by mmitche** [comment]
@bartonjs I restarted the server, that may be why.
@dotnet-bot test Linux arm RElease build

**Comment by bartonjs** [comment]
@dotnet-bot Test Tizen armel Debug Build please

**Comment by bartonjs** [comment]
@dotnet-bot Test Linux x64 Release Build please (System.Composition.Tests on Debian reported incomplete after many hours, but the log say they finished in less than a second)

**Comment by MattGal** [comment]
@bartonjs  did you happen to save anything from that run?  It sounds like it's worth investigating by the FR folks.

Something that has fouled up a lot of test runs recently is tests starting a process, "finishing" then timing out because said process never exits.  I've only seen this in Windows but the same problem might manifest on Linux too.

**Comment by bartonjs** [comment]
@MattGal https://mc.dot.net/#/user/bartonjs/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/b6ac554080266bfc1030b17694803da1102da38c

**Comment by MattGal** [comment]
@bartonjs that's actually not useful since MC only shows the most recent run, but I'll take a quick stab to help out the FR team out.

**Comment by MattGal** [comment]
@bartonjs I reviewed with for a bit with @ChadNedzlek and it's downright spooky.  The migration of the work item's telemetry appears to have just never happened.  For the folks at @dotnet/dncenghot  this is work item # 72945534 in Kusto/SQL.  The test never had any issues, was not retried, but it got lost somewhere.  Very unlikely to be related to @mmitche rebooting a Jenkins thing.  If Chad sees anything actionable & interesting we'll get an issue filed.

**Comment by ChadNedzlek** [comment]
For all the world, it seems like we send this item into an Azure storage Queue, and it just... never got there, but there was no failed report about that happening.  We can add some logging to double check that we actually hit the success path for this message... but none of the error logs triggered, and all the other workitems arrived without issue. And we definitely processed the finished events... from the SQL side, all the normal logs are there, with no abnormalities.

**Pull request merged**